spyder/plugins/editor/utils/autosave.py

```diff
@@ -106,33 +109,78 @@ def stop_autosave_timer(self):
 
-    def single_instance(self):
-        """Return whether Spyder is running in single instance mode."""
-        single_instance = CONF.get('main', 'single_instance')
-        new_instance = self.editor.main.new_instance
-        return single_instance and not new_instance
-
     def do_autosave(self):
         """Instruct current editorstack to autosave files where necessary."""
-        if not self.single_instance():
-            logger.debug('Autosave disabled because not single instance')
-            return
         logger.debug('Autosave triggered')
         stack = self.editor.get_current_editorstack()
         stack.autosave.autosave_all()
         self.start_autosave_timer()
 
-    def try_recover_from_autosave(self):
-        """Offer to recover files from autosave."""
-        if not self.single_instance():
-            self.recover_files_to_open = []
-            return
+    def get_files_to_recover(self):
+        """
+        Get list of files to recover from pid files in autosave dir.
+
+        This returns a tuple `(files_to_recover, pid_files)`. In this tuple,
+        `files_to_recover` is a list of tuples containing the original file
+        names and the corresponding autosave file names, as recorded in the
+        pid files in the autosave directory. Any files in the autosave
+        directory which are not listed in a pid file, are also included, with
+        the original file name set to `None`. The second entry, `pid_files`,
+        is a list with the names of the pid files.
+        """
         autosave_dir = get_conf_path('autosave')
-        autosave_mapping = CONF.get('editor', 'autosave_mapping', {})
-        dialog = RecoveryDialog(autosave_dir, autosave_mapping,
-                                parent=self.editor)
+        if not os.access(autosave_dir, os.R_OK):
+            return [], []
+
+        files_to_recover = []
+        files_mentioned = []
+        pid_files = []
+        non_pid_files = []
+
+        # In Python 3, easier to use os.scandir()
+        for name in os.listdir(autosave_dir):
+            full_name = osp.join(autosave_dir, name)
+            match = re.match(r'pid([0-9]*)\.txt\Z', name)
+            if match:
+                pid_files.append(full_name)
+                logger.debug('Reading pid file: {}'.format(full_name))
+                with open(full_name) as pidfile:
+                    txt = pidfile.read()
+                    txt_as_dict = ast.literal_eval(txt)
+                    files_mentioned += [autosave for (orig, autosave)
+                                        in txt_as_dict.items()]
+                pid = int(match.group(1))
+                if is_spyder_process(pid):
+                    logger.debug('Ignoring files in {}'.format(full_name))
+                else:
+                    files_to_recover += list(txt_as_dict.items())
+            else:
+                non_pid_files.append(full_name)
+
+        # Add all files not mentioned in any pid file
```

Comment:
When does this case happen?

Reply:
It shouldn't, but may happen if the pid file got corrupted. I expanded the comment to clarify.